finat/tensor_product.py

```diff
@@ -162,6 +164,64 @@ def point_evaluation(self, order, point, entity=None):
 
         return self._merge_evaluations(factor_results)
 
+    @property
+    def dual_basis(self):
+        # Outer product the dual bases of the factors
+        qs, pss = zip(*(f.dual_basis for f in self.factors))
+        # TODO: check if this Q_is_identity calculation is correct
+        self.Q_is_identity = all(f.Q_is_identity for f in self.factors)
+        ps = TensorPointSet(pss)
+        tuples_of_q_shape_indices = tuple(gem.indices(len(q.shape)) for q in qs)
+        q_muls = reduce(operator.mul, (q[i] for q, i in zip(qs, tuples_of_q_shape_indices)))
+        # Flatten to get final shape indices
+        q_shape_indices = tuple(index for tuple_of_indices in tuples_of_q_shape_indices for index in tuple_of_indices)
+        Q = gem.ComponentTensor(q_muls, q_shape_indices)
+        return Q, ps
+
+    def dual_evaluation(self, fn):
+        if not hasattr(fn, 'factors'):
+
+            Q, x = self.dual_basis
+
+            #
+            # EVALUATE fn AT x
+            #
+            expr = fn(x)
+
+            #
+            # TENSOR CONTRACT Q WITH expr
+            #
+
+            # NOTE: any shape indices in the expression are because the
+            # expression is tensor valued.
+            # NOTE: here the first num_factors rows of Q are node sets (1 for
+            # each factor)
+            # TODO: work out if there are any other cases where the basis
+            # indices in the shape of the dual basis tensor Q are more than
+            # just the first shape index (e.g. with EnrichedElement)
+            num_factors = total_num_factors(self.factors)
```

Comment:
This feels a little hacky, and seems like it is fragile (I notice that `total_num_factors` has to handle certain elements specially).

Is this interface (returning something with partial shape and partial free indices) the right one for this manipulation and consumers?

I think that at the end of the day we're going to assign something of shape `(num_functionals,)` to an array. But in the TPE case, this is broken down into `(num_functionals_x, num_functionals_y, ...)` ?

Reply:
I think this was resolved in the discussion at the firedrake meeting on 18/08/2021